astrolabe/runner.py

```diff
@@ -163,34 +167,72 @@ def run(self, persist_cluster=False, startup_time=1):
-
-        # Sleep before polling to give Atlas time to update cluster.stateName.
-        sleep(3)
-
-        # Step-4: wait until maintenance completes (cluster is IDLE).
-        selector = BooleanCallablePoller(
-            frequency=self.config.polling_frequency,
-            timeout=self.config.polling_timeout)
-        LOGGER.info("Waiting for cluster maintenance to complete")
-        selector.poll([self], attribute="is_cluster_state", args=("IDLE",),
-                      kwargs={})
-        self.verify_cluster_configuration_matches("final")
-        LOGGER.info("Cluster maintenance complete")
+        for operation in self.spec.operations:
+            if hasattr(operation, 'setClusterConfiguration'):
+                # Step-3: begin maintenance routine.
+                final_config = operation.setClusterConfiguration
+                cluster_config = final_config.clusterConfiguration
+                process_args = final_config.processArgs
+
+                if not cluster_config and not process_args:
+                    raise RuntimeError("invalid maintenance plan")
+
+                if cluster_config:
+                    LOGGER.info("Pushing cluster configuration update")
+                    self.cluster_url.patch(**cluster_config)
+
+                if process_args:
+                    LOGGER.info("Pushing process arguments update")
+                    self.cluster_url.processArgs.patch(**process_args)
+
+                # Sleep before polling to give Atlas time to update cluster.stateName.
+                sleep(3)
+
+                # Step-4: wait until maintenance completes (cluster is IDLE).
+                self.wait_for_idle()
+                self.verify_cluster_configuration_matches(final_config)
+                LOGGER.info("Cluster maintenance complete")
+                
+            if hasattr(operation, 'testFailover'):
+                self.cluster_url['restartPrimaries'].post()
+                
+                self.wait_for_idle()
+                
+            if hasattr(operation, 'sleep'):
+                _time.sleep(operation['sleep'])
+                
+            if hasattr(operation, 'waitForIdle'):
+                self.wait_for_idle()
+                
+            if hasattr(operation, 'restartVms'):
+                url = "/api/private/nds/groups/%s/clusters/%s/reboot" % (self.project.id, self.cluster_name)
+                self.admin_client.request('POST', url)
+                
+                sleep(5)
+                self.wait_for_idle()
+                
+            if hasattr(operation, 'assertPrimaryRegion'):
+                region = operation['assertPrimaryRegion']
+                
+                cluster_config = self.cluster_url.get().data
+                deadline = monotonic() + 90
+                
+                while True:
+                    mc = MongoClient(cluster_config['connectionStrings']['standard'], username='atlasuser', password='mypassword123')
```

Comment:
Why does this need a new MongoClient? Why can we not use one of the clients attached to the context for this operation?

Reply:
I see two other clients in the codebase but neither of them appears to be attached to any contexts, which client did you have in mind exactly?